### DISCUSSION TITLE
Get PC username and PC name

### DISCUSSION THREAD

**Discussion by AppetiteTeam** [halildurmus/win32#377] [Q&A] [answered]
### I don't understand how to get username and PC name. Tried like this:
`    final localeName = calloc<Utf16>();`
    `final user = GetUserDefaultLocaleName(localeName, LOCALE_USER_DEFAULT);`

### VS Code shows that there are no errors, but when compiling it gives this:

: Error: Expected type 'Utf16' to be a valid and instantiated subtype of 'NativeType'.
package:pc_remote/helpers/media.dart:31
- 'Utf16' is from 'package:ffi/src/utf16.dart' ('/[C:/FlutterSDK/.pub-cache/hosted/pub.dartlang.org/ffi-1.1.2/lib/src/utf16.dart]()').
package:ffi/src/utf16.dart:1

**Comment by timsneath** [comment, marked as answer]
That's correct. `Utf16` is not a type that can be allocated (it's an opaque type that exists primarily to enable extension methods to convert back to a string). Instead you allocate a native type like `Uint16` (which represents 2 bytes of storage, enough for a single Unicode character) and then cast it to a `Utf16`. And of course, you need enough of them for your buffer (use the optional parameter). For example:

```dart
final buffer = calloc<Uint16>(256).cast<Utf16>();
```

To make it easier, win32 offers a helper method that takes care of these details for you:

```dart
final buffer = wsalloc(256);
```

Saving you a lot of trouble, this example shows how to get the username and machine name:
https://github.com/timsneath/win32/blob/main/example/sysinfo.dart
